nifi-nar-bundles/nifi-kite-bundle/nifi-kite-processors/src/main/java/org/apache/nifi/processors/kite/InferAvroSchema.java

```diff
@@ -0,0 +1,454 @@
+                                out.write(avroSchema.get().getBytes());
+                            }
+                        });
+                        avroSchemaFF = session.putAttribute(avroSchemaFF, CoreAttributes.MIME_TYPE.key(), AVRO_MIME_TYPE);
+                        break;
+                    default:
+                        break;
+                }
+
+                //Transfer the sessions.
+                avroSchemaFF = session.putAttribute(avroSchemaFF, CoreAttributes.FILENAME.key(), (original.getAttribute(CoreAttributes.FILENAME.key()) + AVRO_FILE_EXTENSION));
+                session.transfer(avroSchemaFF, REL_SUCCESS);
+            } else {
+                //If the avroSchema is null then the content type is unknown and therefore unsupported
+                session.transfer(session.clone(original), REL_UNSUPPORTED_CONTENT);
+            }
+
+            session.transfer(original, REL_ORIGINAL);
+
+        } catch (Exception ex) {
+            getLogger().error("Failed to infer Avro schema! {}", ex);
+            session.transfer(original, REL_FAILURE);
+        }
+    }
+
+
+    /**
+     * Infers the Avro schema from the input Flowfile content. To infer an Avro schema for CSV content a header line is
+     * required. You can configure the processor to pull that header line from the first line of the CSV data if it is
+     * present OR you can manually supply the desired header line as a property value.
+     *
+     * @param inputFlowFile
+     *  The original input FlowFile containing the CSV content as it entered this processor.
+     *
+     * @param context
+     *  ProcessContext to pull processor configurations.
+     *
+     * @param session
+     *  ProcessSession to transfer FlowFiles
+     */
+    private AtomicReference<String> inferAvroSchemaFromCSV(final FlowFile inputFlowFile, final ProcessContext context, final ProcessSession session) {
+
+        //Determines the header line either from the property input or the first line of the delimited file.
+        final AtomicReference<String> header = new AtomicReference<>();
+        final AtomicReference<Boolean> hasHeader = new AtomicReference<>();
+
+        if (context.getProperty(GET_CSV_HEADER_DEFINITION_FROM_INPUT).asBoolean() == Boolean.TRUE) {
+            //Read the first line of the file to get the header value.
+            session.read(inputFlowFile, new InputStreamCallback() {
+                @Override
+                public void process(InputStream in) throws IOException {
+                    BufferedReader br = new BufferedReader(new InputStreamReader(in));
+                    header.set(br.readLine());
+                    hasHeader.set(Boolean.TRUE);
+                    br.close();
+                }
+            });
+            hasHeader.set(Boolean.TRUE);
+        } else {
+            header.set(context.getProperty(CSV_HEADER_DEFINITION).evaluateAttributeExpressions().getValue());
```

Comment:
this should evaluate expressions against the FlowFile: ...evaluateAttributeExpressions(inputFlowFile).getValue()